File: docs-examples/example-java/src/test/java/io/micronaut/rabbitmq/docs/serdes/ProductInfoSerDesSpec.java
package io.micronaut.rabbitmq.docs.serdes;

import io.micronaut.context.annotation.Property;
import io.micronaut.test.extensions.junit5.annotation.MicronautTest;
import io.micronaut.test.support.TestPropertyProvider;
import org.junit.jupiter.api.Test;

import static java.util.concurrent.TimeUnit.SECONDS;
import static org.awaitility.Awaitility.await;
import io.micronaut.rabbitmq.testcontainers.RabbitMQ;
import org.junit.jupiter.api.TestInstance;

import java.util.Map;

@MicronautTest
@Property(name = "spec.name", value = "ProductInfoSerDesSpec")
@TestInstance(TestInstance.Lifecycle.PER_CLASS)
class ProductInfoSerDesSpec implements TestPropertyProvider {
    @Override
    public Map<String, String> getProperties() {
        return RabbitMQ.getProperties();
    }

    @Test
    void testUsingACustomSerDes(ProductClient productClient, ProductListener productListener) {
// tag::producer[]
        productClient.send(new ProductInfo("small", 10L, true));
        productClient.send(new ProductInfo("medium", 20L, true));
        productClient.send(new ProductInfo(null, 30L, false));
// end::producer[]


        await().atMost(60, SECONDS).until(() ->
                productListener.messages.size() == 3 &&
                productListener.messages.stream().anyMatch(pi -> pi.getCount() == 10L) &&
                productListener.messages.stream().anyMatch(pi -> pi.getCount() == 20L) &&
                productListener.messages.stream().anyMatch(pi -> pi.getCount() == 30L)
        );
    }
}

File: docs-examples/example-python/src/test/python/micronaut/rabbitmq/docs/serdes/ProductInfoSerDesSpec.py
from typing import Annotated

from jakarta.inject import Inject
from micronaut.context.annotation import Property
from micronaut.rabbitmq.docs.Await import Await
from micronaut.test.extensions.junit5.annotation import MicronautTest
from org.junit.jupiter.api import Test

from .ProductClient import ProductClient
from .ProductInfo import ProductInfo
from .ProductListener import ProductListener


@MicronautTest(environments=["rabbitmq"])
@Property(name="spec.name", value="ProductInfoSerDesSpec")
class ProductInfoSerDesSpec:
    product_client: Annotated[ProductClient, Inject]
    product_listener: Annotated[ProductListener, Inject]

    @Test
    def test_using_a_custom_serdes(self) -> None:
# tag::producer[]
        self.product_client.send(ProductInfo("small", 10, True))
        self.product_client.send(ProductInfo("medium", 20, True))
        self.product_client.send(ProductInfo(None, 30, False))
# end::producer[]

        messages = self.product_listener.messages
        Await.until(lambda: len(messages) == 3
                    and any(pi.count == 10 for pi in messages)
                    and any(pi.count == 20 for pi in messages)
                    and any(pi.count == 30 for pi in messages))
